File: aoc2022/src/day8/day8a.rs
use std::fs::read_to_string;

pub fn read_input_example() -> String {
    read_to_string("data/day8/example.txt").unwrap()
}

pub fn read_input() -> String {
    read_to_string("data/day8/a.txt").unwrap()
}

pub fn solve(input: String) -> i32 {
    let grid = parse_grid(input);
    let mut count = 0i32;
    for (y, row) in grid.iter().enumerate() {
        for (x, tree) in row.iter().enumerate() {
            if x == 0 || x == row.len() - 1 {
                count += 1;
                continue;
            }
            if y == 0 || y == grid.len() - 1 {
                count += 1;
                continue;
            }
            let north = (0..y)
                .map(|dy| (x, dy))
                .map(|(dx, dy)| grid[dy][dx])
                .all(|t| *tree > t);
            let east = (x + 1..row.len())
                .map(|dx| (dx, y))
                .map(|(dx, dy)| grid[dy][dx])
                .all(|t| *tree > t);
            let south = (y + 1..grid.len())
                .map(|dy| (x, dy))
                .map(|(dx, dy)| grid[dy][dx])
                .all(|t| *tree > t);
            let west = (0..x)
                .map(|dx| (dx, y))
                .map(|(dx, dy)| grid[dy][dx])
                .all(|t| *tree > t);
            if north || east || south || west {
                count += 1;
            }
        }
    }
    count
}

pub fn parse_grid(input: String) -> Vec<Vec<i32>> {
    input
        .trim()
        .split("\n")
        .map(|line| {
            line.chars()
                .map(|c| c.to_digit(10).unwrap() as i32)
                .collect()
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn should_solve_example() {
        let input = read_input_example();
        let result = solve(input);
        assert_eq!(result, 21);
    }

    #[test]
    fn should_solve() {
        let input = read_input();
        let result = solve(input);
        assert_eq!(result, 1803);
    }
}

File: aoc2023/src/day8/day8a.ts
import { readFileSync } from 'fs'

export function readInput(): string {
    return readFileSync('data/day8.txt').toString().trim()
}

export function solve(input: string): number {
    const inp = input.split('\n\n')
    const moves = inp[0].split('')
    const map = new Map(inp[1].split('\n').map(l => {
        const v = l.split('=')[1].split(', ')
        return [l.split(' ')[0], [v[0].slice(2), v[1].slice(0, -1)]]
    }))
    let i = 0
    let pos = 'AAA'
    while (true) {
        const m = moves[i % moves.length]
        pos = map.get(pos)![m === 'L' ? 0 : 1]
        if (pos === 'ZZZ') {
            return i + 1
        }
        i++
    }
}
